contracts/expiring_plugin.algo.ts

```diff
@@ -0,0 +1,106 @@
+import { Contract } from '@algorandfoundation/tealscript';
+
+/**
+ * ExpiringPluginFactory is a factory for dynamic general purpose expiring plugins.
+ * They allow you to delegate a specific address the ability 
+ * to call a single method on a specific other contract for a limited time.
+ */
+export class ExpiringPluginFactory extends Contract {
+    programVersion = 10;
+
+    /**
+     * Mint a new expiring plugin with a specific set of rules
+     * @param duration 
+     * @param authAddr 
+     * @param allowedApp 
+     * @param method 
+     */
+    newExpiringPlugin(duration: uint64, authAddr: Address, allowedApp: Application, method: bytes): void {
+        sendMethodCall<[uint64, Address, Application, bytes], void>({
+            approvalProgram: ExpiringPlugin.approvalProgram(),
+            clearStateProgram: ExpiringPlugin.clearProgram(),
+            name: 'createApplication',
+            methodArgs: [duration, authAddr, allowedApp, method], 
+        });
+    }
+}
+
+
+/**
+ * ExpiringPlugin is a dynamic general purpose expiring plugin.
+ * It allows you to delegate a specific address the ability 
+ * to call a single method on a specific other contract for a limited time.
+ */
+export class ExpiringPlugin extends Contract {
+    programVersion = 10;
+
+    /**
+     * The time at which this contract was created
+     */
+    createdAt = GlobalStateKey<uint64>();
+
+    /**
+     * The duration is valid for
+     */
+    duration = GlobalStateKey<uint64>();
+
+    /**
+     * The address that is authorized to call the method
+     */
+    authAddr = GlobalStateKey<Address>();
+
+    /**
+     * The application that is authorized to be called
+     */
+    allowedApp = GlobalStateKey<Application>();
+
+    /**
+     * The method that is authorized to be called
+     */
+    method = GlobalStateKey<bytes>();
+
+    /**
+     * set the all state values at mint
+     * @param duration 
+     * @param authAddr 
+     * @param allowedApp 
+     * @param method 
+     */
+    createApplication(duration: uint64, authAddr: Address, allowedApp: Application, method: bytes): void {
+        this.createdAt.value = globals.latestTimestamp;
+        this.duration.value = duration;
+        this.authAddr.value = authAddr;
+        this.allowedApp.value = allowedApp;
+        this.method.value = method;
+    }
+
+    /**
+     * verifyAppCall ensures the external app call being made
+     * is valid and the contract hasn't expired
+     */
+    verify(sender: Account, appCall: AppCallTxn): void {
```

Comment:
Rather than depending on external method call validation, I feel like it would make more sense to make this a private method and allow apps to inherit from this contract to perform an expiration check.

Reply:
im trying to think how this would be managed if it wasn't an external contract. If we used inheritance here wouldn't it be difficult to make this something that could be used repetitively? I was thinking you may have users that regularly give a good amount of apps permission to call other methods on your behalf

Reply:
I envision factories being a separate type of application stored in box storage in the abstracted account. Factories would allow for safe creation of plugins with defined functionality for any app. It would basically replace the need for templated apps. For example, if an app wants to do a recurring payment, it would ask you to mint a subscription plugin from the subscription factory with specific parameters. Factories would be called by the user and the factory would have permission to add to the plugins

Reply:
okay so if were bring factories into the abstracted contract itself we get the benefit of
 - known list of safe or atleast non-malicious factory apps
 - the ability for the factory to add the child contract to the plugin list in an itxn
 
 i wasn't aware we could place contracts in box storage, still piecing this together in my head

Reply:
ahh i see, so really just extending this from an entirely separate contract to something the contract can manage a list of factories for kinda like we mentioned before about internal factories.

Ill make the adjustments